Plugins: Fix Renew Now action, take 2

sirbrillig opened the pull request:
#### Changes proposed in this Pull Request

This is modified reimplementation of https://github.com/Automattic/wp-calypso/pull/62169 which had to be reverted in https://github.com/Automattic/wp-calypso/pull/62438.

In addition, this refactors the `getRenewalItemFromProduct()` function so that it uses early returns instead of a reassigned variable to help prevent such issues in the future. (To do this in a DRY way, this creates a private sub-function to create the initial item.) In order to preserve the behavior of the original version of the function, this also reverses the order of the conditions, since an early return is the opposite of a variable being overwritten.

#### Testing instructions

This adds unit tests to prevent the regression that caused the revert, but tests only test what we expect to find.

See bug report for https://github.com/Automattic/wp-calypso/issues/62128 to see if this fixes that bug.

Verify that clicking to renew a domain product still works.

sirbrillig committed:
Refactor getRenewalItemFromProduct to use early returns

sirbrillig committed:
Re-add isRenewable check for plugins

epeicher commented:
Thanks for your PR @sirbrillig! The change looks good to me and it works as expected with regards to the issue https://github.com/Automattic/wp-calypso/issues/62128

About
> check not just for the isRenewable property but also to make sure that the product is a plugin somehow

I am not aware of a way to know if a product is a plugin. I have asked Bespin team and I will post something here if I find an alternative solution for that.

epeicher committed:
Check if a product is from Marketplace when renewing

epeicher reviewed (approved):
I tested the related issue https://github.com/Automattic/wp-calypso/pull/62440 and works as expected.

Changes LGTM!

sirbrillig commented:
Thank you! Before I merge this I'm just going to check the logic of a few of the conditions to make sure they didn't rely on the previous "overwriting" behavior. Particularly I want to make sure that `isPlan` and `isJetpackProduct` don't overlap.

sirbrillig committed:
Reverse the order some of the conditions to preserve old behavior

sirbrillig commented on client/lib/purchases/index.js:
```diff
@@ -138,6 +138,7 @@ function getSubscriptionEndDate( purchase ) {
 function handleRenewNowClick( purchase, siteSlug, options = {} ) {
 	const renewItem = getRenewalItemFromProduct( purchase, {
 		domain: purchase.meta,
+		isMarketplaceProduct: options.isMarketplaceProduct,
```
The only thing that worries me about this approach is that there are other places in the codebase that call `getRenewalItemFromProduct()` and will not have this option provided. 🤔

sirbrillig replied:
I've removed this flag for reasons I explained more about below.

sirbrillig committed:
Add unit tests to prevent the previous regression

ramynasr commented on client/me/purchases/manage-purchase/index.jsx:
```diff
@@ -180,8 +180,11 @@ class ManagePurchase extends Component {
-		const options = redirectTo ? { redirectTo } : undefined;
+		const { purchase, siteSlug, redirectTo, productsList } = this.props;
+
+		const options = { redirectTo };
```
This seems to change the behavior of the code before, which checks for `redirectTo` before it gets added. I don't know if that would make any difference. But would that be a safer alternative?

```
const options = redirectTo ? { redirectTo } : {};
```

rafaelgallani replied:
For this case, I think it wouldn't be a problem since `options` has a default value of `{}` in `handleRenewNowClick`:

https://github.com/Automattic/wp-calypso/blob/78d74cca5c6f8ac75d667f435ebfbfb3c5281427/client/lib/purchases/index.js#L138

Also, there's a check for the `redirectTo` prop later on:

https://github.com/Automattic/wp-calypso/blob/78d74cca5c6f8ac75d667f435ebfbfb3c5281427/client/lib/purchases/index.js#L163

sirbrillig replied:
Good point. I _think_ it's safe, but for other reasons I've reverted this change anyway,

sirbrillig committed:
Add test for isRenewable

sirbrillig committed:
Remove isMarketplaceProduct option in createRenewalCartItemFromProduct

rafaelgallani reviewed (commented):
Nice job! I just tested and confirmed that renewing domains works properly.

sirbrillig commented:
After considering this some more, I think it's safer to allow `getRenewalItemFromProduct()` to return a renewal item for any `isRenewable` product (as was implemented in the now-reverted PR) without also verifying it to be a plugin. Here's my reasoning.

`getRenewalItemFromProduct()` is a fairly pure function, and it is called in many place in the codebase, so we really would need a consistent means to determine if a product is a plugin without network calls. From the research I've done there doesn't appear to be any way to identify such products currently from a purchase object alone, and I can find no hard-coded list of plugin slugs like the ones we use for other product types. Ideally, we'd modify the endpoint that generates this object to add a flag for plugins (like the current `is_domain_registration` flag), but since this is a bug fix for a live feature that feels out of scope for now.

I've reverted the additional code added that set the `isMarketplaceProduct` option and added a unit test for the `isRenewable` behavior. We can always revisit this all later.
